[IMP] product: optimize create_variant_ids by doing batches

seb-odoo opened the pull request:
The biggest impact of this change is when each template has only very few
variants to generate and there are a lot of such templates.
And when few templates generate a lot of variants this is potentially just a bit
better than before, and not definitely not worse.

The opposite change was done in commit f8dad1b
due to an issue with unlink but since the unlink is changed to use dichotomy now
we have the best of both solutions.

For example when creating 10000 products, few with a lot of variants (max 1000)
and many with few variants (or only one), the time it takes goes down from at
least 10min (the attempt was canceled before the end) to only 25s.

Part of task-1918881

seb-odoo commented:
@JKE-be here's probably my PR with the biggest impact on batch variant creation

JKE-be commented on addons/product/models/product.py:
```diff
@@ -422,6 +422,40 @@ def unlink(self):
+            # doesn't fail.
+            if len(self) > 1:
+                self[:len(self) // 2]._unlink_or_archive()
+                self[len(self) // 2:]._unlink_or_archive()
```
@antonylesuisse 

your pov ?

seb-odoo replied:
For the following test case, realistic enough:

- create a template
- give it 3 attributes, 10 values each
- -> that generates 1000 variants
- add the first variant to an order to test it
- not happy with the config, remove one of the attributes from the template
- -> that removes the 1000 variants

This is slow because of the `savepoint` used to archive if `unlink` fails.

With a loop (one by one) it takes 5min.
With dichotomy it takes 14s.

I added this result on the commit message.

seb-odoo commented:
@JKE-be I confirmed with AL the dichotomy is acceptable considering the significant gain on the favorable use case.

Also rebased PR and removed the api.multi that I added.

Good to go?

JKE-be commented:
@robodoo r+

robodoo commented:
Merged, thanks!
